File: 01_Language/01_Functions/python/array_reduce.py
# coding: utf-8

import functools


def array_reduce(arr, cb, *args):
    if len(args) > 0:
        return functools.reduce(cb, arr, args[0])
    else:
        return functools.reduce(cb, arr)


if __name__ == '__main__':
    l = [1, 2, 3, 4, 5]
    print(array_reduce(l, lambda x, y: x + y))
    print(array_reduce(l, lambda x, y: x * y, 10))
    print(array_reduce([], lambda x, y: x + y, "No data to reduce"))

File: 01_Language/01_Functions/js/array_reduce.js
function array_reduce(arr, cb) {
    if (arguments.length > 2) {
        return arr.reduce(cb, arguments[2])
    } else {
        return arr.reduce(cb)
    }
}

let list = [1, 2, 3, 4, 5]
console.log(array_reduce(list, function (prev, cur, index, arr) {
    return prev + cur
}));
console.log(array_reduce(list, function (prev, cur, index, arr) {
    return prev * cur
}, 10));
console.log(array_reduce([], function (prev, cur, index, arr) {
    return prev + cur
}, "No data to reduce"));
